mesa/agent.py

```diff
@@ -37,14 +38,28 @@ class Agent:
         self.pos: Position | None = None
     """
 
-    def __init__(self, unique_id: int, model: Model) -> None:
+    _ids = itertools.count()
+
+    def __init__(self, *args, **kwargs) -> None:
         """
         Create a new agent.
 
         Args:
             unique_id (int): A unique identifier for this agent.
             model (Model): The model instance in which the agent exists.
         """
+        if len(args) == 1:
+            model = args[0]
+            unique_id = next(self._ids)
+        else:
+            warnings.warn(
+                "unique ids are assigned automatically in MESA 3",
+                DeprecationWarning,
+                stacklevel=2,
+            )
+            unique_id, model, *args = args
```

Comment:
Now you’re allowing users to assign their own unique_id right, isn’t that exactly what we don’t want?

It does make the “break” smaller though. But if we’re going to replace the warning before 3.0, what’s the point?

Reply:
Like normally I would like this: first a PR that introduces a warning, and then a few minor releases later we remove the deprecated behavior.

But now we’re jumping to 3.0, and PR this allows two different things at once. We can’t trust on the unique id being unique.

This might be useful as a 2.4 backport to warn for new behavior.

Reply:
> Like normally I would like this: first a PR that introduces a warning, and then a few minor releases later we remove the deprecated behavior.

You equate a single PR with a release. My suggestion would be first to do this PR, then fix all deprecation warnings in the tests as a PR, and then remove this `else` statement in a PR. Once these 3 atomic PRs are done, this entire feature is ready for a next release. 

> But now we’re jumping to 3.0, and PR this allows two different things at once. We can’t trust on the unique id being unique.

I am sorry, but I just don't follow. Yes, this code allows the old code to still function but with a deprecation warning that allows us to clean up the rest of the code base. Once the rest of the code base is cleaned up, this can also be removed. 

I don't understand what any off this has to do with trusting unique_id to be truly unique. The current 2.x code also cannot guarantee that.

Reply:
Most of the errors are not in the structure of the arguments, but in the values themselves. A lot of our tests demanded very specific `unique_id` values. I fixed those tests and that’s #2226.